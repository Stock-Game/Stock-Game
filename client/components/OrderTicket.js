import React, { useState, useEffect } from 'react';
import axios from 'axios';

export default function OrderTicket() {
  // set default state
  const [ticker, setTicker] = useState('');
  const [order, setOrder] = useState({
    ticker: '',
    price: '',
    quantity: 0,
    total: '',
    action: 'buy',
  });

  // onchange update
  const handleTickerChange = (e) => {
    setTicker(e.target.value);
  };

  const handleSubmit = (e) => {
    // get request to server with ticker in req.params
    e.preventDefault();
    axiosFetchData();
    axiosSyncData();
    setTicker('');
  };
  //fetch from orderTicket
  const axiosFetchData = async () => {
    await axios
      .get(`http://localhost:4000/orderTicket`, {
        params: { ticker: ticker.toUpperCase() },
      })
      .then((res) => {
        setOrder({
          ...order,
          ticker: res.data.symbol,
          price: res.data.price,
        });
      })
      .catch((err) => console.log('error: ', err));
  };
  // get orderTicket and sync with portfolio
  const axiosSyncData = async () => {
    await axios
      .get('http://localhost:4000/portfolio/sync')
      .catch((err) => console.log(err));
  };

  const submitOrder = async () => {
    const postData = {
      ticker: order.ticker,
      priceBought: order.price,
      shares: order.quantity,
      totalCost: order.total,
    };
    if (order.action === 'buy') {
      // buy route = post
      await axios
        .post(`http://localhost:4000/portfolio`, postData)
        .catch((err) => console.log('error: ', err));
    } else if (order.action === 'sell') {
      // sell route = delete
      await axios
        .delete(`http://localhost:4000/portfolio`, { data: postData })
        .catch((err) => console.log('error: ', err));
    }
  };

  return (
    <>
      <div>
        <div className='orderBox'>
          <form onSubmit={handleSubmit}>
            <div className='row2'>
              <textarea
                placeholder='Enter a stock ticker'
                rows='1'
                value={ticker}
                onChange={handleTickerChange}
              />
              <button disabled={ticker.length === 0}>Search</button>
            </div>
          </form>
          <table>
            <tbody>
              <tr>
                <td>
                  <strong>Ticker</strong>
                </td>
                <td>{order.ticker}</td>
              </tr>
              <tr>
                <td>
                  <strong>Price</strong>
                </td>
                <td>{order.price}</td>
              </tr>
              <tr>
                <td>
                  <strong>Quantity</strong>
                </td>
                <td>
                  <input
                    type='number'
                    value={order.quantity || ''}
                    onChange={(e) => {
                      setOrder({
                        ...order,
                        quantity: Math.max(e.target.value, 1),
                        total: order.quantity * order.price,
                      });
                    }}
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <strong>Total Transaction Amount</strong>
                </td>
                <td>
                  {Math.ceil((order.price * order.quantity) * 100) / 100 || ''}
                  {/* {Math.ceil(((order.price * order.quantity) * 100) / 100) || ''} */}
                </td>
              </tr>
              <tr>
                <td>
                  <strong>Action</strong>
                </td>
                <td>
                  {/* create a select */}
                  <select
                    value={order.action}
                    onChange={(e) => {
                      setOrder({
                        ...order,
                        action: e.target.value,
                      });
                    }}
                  >
                    <option value='buy'>Buy</option>
                    <option value='sell'>Sell</option>
                  </select>
                </td>
              </tr>
              <tr>
                <td></td>
                <td>
                  <button
                    onClick={() => {
                      setOrder({
                        ticker: '',
                        price: '',
                        quantity: 0,
                        total: '',
                        action: '',
                      });
                      console.log(
                        `Placing a ${order.action} order for ${order.quantity} shares of ${order.ticker} at ${order.price}`
                      );
                      // reload on submit before and after submitOrder function
                      location.reload();
                      submitOrder();
                      location.reload();
                    }}
                    disabled={order.total === ''}
                  >
                    Submit
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </>
  );
}
