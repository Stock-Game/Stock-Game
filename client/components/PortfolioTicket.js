import React, { useState, useEffect } from 'react';

export default function PortfolioTicket(props) {
  console.log(props.priceInfo);
  return (
    //  ** the names of the props.ticketInfo properties will likely be changed **
    // all styling will likely be moved to a css file, it's just not being served yet (i think)
    <>
      <td className='ticketcells'>{props.ticketInfo.ticker}</td>
      <td id='purchasePrice' className='ticketcells'>
        ${Math.floor(props.priceInfo.price).toLocaleString('en-US')}
      </td>
      <td className='ticketcells'>{props.ticketInfo.shares}</td>
      <td className='ticketcells'>
        $
        {Math.floor(
          props.priceInfo.price * props.ticketInfo.shares
        ).toLocaleString('en-US')}
      </td>
      <td className='ticketcells'>
        $
        {Math.floor(
          (props.priceInfo.price - props.ticketInfo.priceBought) *
            props.ticketInfo.shares
        ).toLocaleString('en-US')}
      </td>
    </>
  );
}
