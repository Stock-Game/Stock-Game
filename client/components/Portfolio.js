import React, { useState, useEffect } from 'react';
import PortfolioTicket from './PortfolioTicket';
import BalancesDisplay from './BalancesDisplay';
// make a call to the database and grab all current positions
// use those positions to populate portfolio

// probably will loop thru all positions and render an array of PortfolioTicket components
// and pass all position info down to PortfolioTicket components thru props

export default function Portfolio() {
  const testBalance = {
    totalVal: 100000,
    cash: 20000,
    portfolioBal: 120000,
  };

  const [positionArr, setPositionArr] = useState([]);

  const [prices, setPrices] = useState([]);

  const [balance, setBalance] = useState({
    totalVal: 0,
    cash: 200000,
    portfolioBal: 100000,
  });

  useEffect(() => {
    getPortfolioData();
  }, []);

  useEffect(() => {
    getBalanceData();
  }, [positionArr]);

  const getPortfolioData = () => {
    fetch('http://localhost:4000/portfolio', {
      method: 'GET',
    })
      .then((res) => res.json())
      .then((data) => {
        console.log(data);
        setPositionArr(data[0]);
        setPrices(data[1]);
        console.log('');
      });
  };

  const syncPrices = () => {
    fetch('http://localhost:4000/portfolio/sync', {
      method: 'GET',
    })
      .then((res) => res.json())
      .then((data) => {
        console.log('DATA FROM SYNC: ', data);
        setPrices(data[0]);
      });
  };

  const getBalanceData = () => {
    let priceBought = 0;
    let totalPortVal = 0;
    positionArr.forEach((el) => {
      const thisPriceInfo = prices.find((element) => {
        return el.ticker === element.ticker;
      });
      console.log('this price info: ', thisPriceInfo);
      priceBought += el.priceBought * el.shares;
      totalPortVal += thisPriceInfo.price * el.shares;
    });
    const cash = 200000 - priceBought;
    console.log('price bought: ', priceBought);
    setBalance({
      totalVal: totalPortVal,
      cash: cash - priceBought,
      portfolioBal: cash + totalPortVal,
    });
  };

  return (
    // all styling will likely be moved to a css file, it's just not being served yet (i think)
    <>
      <div
        id='portfoliocontainer'
      >
        <button onClick={syncPrices}>SYNC</button>
        <table id='portfoliotable'>
          <thead id='titlehead'>
            <tr id='titlerow'>
              <td className='ticketkey'>Ticker</td>
              <td className='ticketkey'>Price</td>
              <td className='ticketkey' id='quantitystyle'>
                Quantity
              </td>
              <td className='ticketkey' id='valuestyle'>
                Total Value
              </td>
              <td className='ticketkey'>Profit/Loss</td>
            </tr>
          </thead>
          <tbody id='porttablebody'>
            {positionArr.map((el) => (
              <tr id='portfolioticket'>
                <PortfolioTicket
                  ticketInfo={el}
                  priceInfo={prices.find((element) => {
                    return element.ticker === el.ticker;
                  })}
                  key={'key ' + el.ticker}
                />
              </tr>
            ))}
          </tbody>
        </table>
      </div>
      <BalancesDisplay balanceInfo={balance} />
    </>
  );
}
