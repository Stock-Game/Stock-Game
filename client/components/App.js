import React, { useState, useEffect } from 'react';
import ReactDOM from 'react-dom/client';
import { Routes, Route, BrowserRouter, Switch } from 'react-router-dom';
import Portfolio from './Portfolio';
import OrderTicket from './OrderTicket';
import NewsDisplay from './NewsDisplay';
import Signup from './Signup';
import HomePage from './HomePage';

export default function App() {
  return (
    <div className='app'>
      <span>
        <img
          height='150px'
          src='https://justkillingti.me/wp-content/2016/09/trading-places-aykroyd-murphy.png'
        />
        <h1>Trading Simulator</h1>
      </span>
      <br />
      <h1>Buy or sell a stock:</h1>

      <Routes>
        <Route exact path='/logup/signup' element={<Signup />} />
        <Route index element={<HomePage />} />
        {/* <Route exact path='/' component={HomePage} /> */}
      </Routes>
    </div>
  );

  // return (
  //   <>
  //     <div className='row'>
  //       {/* <img
  //         height={100}
  //         src={
  //           'https://factanimal.com/wp-content/uploads/2022/03/red-lipped-batfish-facts.jpg'
  //         }
  //       /> */}
  //       <h1>Trading Simulator</h1>
  //     </div>
  //     <br />
  //     <h1>Buy or sell a stock:</h1>
  //     <div className='row2'>
  //       <OrderTicket />
  //       <Portfolio />
  //     </div>
  //       <NewsDisplay />
  //   </>
  // );
}
