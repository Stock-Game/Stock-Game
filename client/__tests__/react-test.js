import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen, waitFor } from '@testing-library/react';

import App from '../components/App';
import Portfolio from '../components/Portfolio';
import PortfolioTicket from '../components/PortfolioTicket';
import OrderTicket from '../components/OrderTicket';

describe('Order Ticket', () => {
  let ticket;

  beforeEach(() => {
    ticket = render(<OrderTicket />);
  });

  test('should display order ticket', () => {
    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(screen.getByText('Ticker')).toBeInTheDocument();
    expect(screen.getByText('Price')).toBeInTheDocument();
    expect(screen.getByText('Quantity')).toBeInTheDocument();
    expect(screen.getByText('Total Transaction Amount')).toBeInTheDocument();
    expect(screen.getByText('Action')).toBeInTheDocument();
    expect(screen.getByText('Buy')).toBeInTheDocument();
    expect(screen.getByText('Sell')).toBeInTheDocument();
  });

  test('contains buttons labeled Search and Submit', () => {
    // expect(ticket.getByText("Submit", { selector: "button" }));

    const buttons = screen.getAllByRole('button');
    expect(buttons.length).toBe(2);
    expect(buttons[0]).toHaveTextContent('Search');
    expect(buttons[1]).toHaveTextContent('Submit');
  });
  test('The functions passed down should be invoked on click', () => {
    const myMock1 = jest.fn();

    userEvent.click(screen.getByText('Search'));
    expect(myMock1).toHaveBeenCalled();
  });

  test('Transaction amount no more than 2 decimals', () => {
    const order = {
      price: 1.3333,
      quantity: 5,
    };

    const total = Math.ceil(order.price * order.quantity * 100) / 100;
    expect(total).toBe(6.67);
  });

  test('Snapshot test', () => {
    expect(OrderTicket).toMatchSnapshot();
  });
});
