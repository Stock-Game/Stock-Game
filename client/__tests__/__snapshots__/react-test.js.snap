// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Ticket Snapshot test 1`] = `[Function]`;
